### DISCUSSION TITLE
supabase multi-tenancy

### DISCUSSION THREAD

**Discussion by mrcasablr** [supabase/supabase#5071] [General]
Hello!

I looked over previous multi-tenancy questions and I understand it is possible via RLS. We plan to use the "pool" model same db and schema for multi-tenancy to simplify architecture and scale.

Here is my use case -
Build a platform for small-medium manufacturing for employee management.  

There are multiple manufacturing organizations (org_id)
Each organization has multiple factories (factory_id)
Each factory has multiple departments (dept_id)
Each department has employees (emp_id)

When we talk multi-tenancy, how do we ensure that an employee requesting access to certain departmental data, only gets data from that specific org/factory/department using RLS

Thanks in advance!

**Comment by GaryAustin1** [comment]
A very simple way would be to have a role table per user (have to populate it somehow though so need interface to handle that).

The role table would have org,fact,dept,emp roles with id (or maybe even array of id's per type if someone might access only 2 of 3 factories versus the whole organization)

Your tables then have all higher level ids in them.  So a department has  dept, fact and org ids.

The RLS can then say 
1) if org id in user's role org id array, then this grant access to the org and all factories, departments, employees with that org id
2) else check fact id in user's role fact id array, then grant access to that factory and all departments, employees with that fact id
etc.

You can add additional roles for reading/updating other functions on different tables as needed.  I suspect in the end though special cases of what columns they can see, what you can change, etc. can make things very complex unless your applications is simple.  Views for different roles might also help in controlling parts of tables they can see.  

I would assume there are many examples online if you search postgres multi tenant with ideas and pitfalls of different approaches.
